npm/debugTools/dataplane/tupleProcessor.go

```diff
@@ -0,0 +1,431 @@
+	}
+
+	hitRules, err := getHitRules(srcPod, dstPod, allRules, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get hit rules : %w", err)
+	}
+
+	if len(hitRules) == 0 {
+		// either no hit rules or no rules at all. Both cases allow all traffic
+		hitRules = append(hitRules, &pb.RuleResponse{Allowed: true})
+	}
+
+	ruleResListJSON := make([][]byte, 0)
+	m := protojson.MarshalOptions{
+		Indent: "	",
+		EmitUnpopulated: true,
+	}
+	for _, rule := range hitRules {
+		ruleJSON, err := m.Marshal(rule) // pretty print
+		if err != nil {
+			return nil, nil, fmt.Errorf("error occurred during marshalling : %w", err)
+		}
+		ruleResListJSON = append(ruleResListJSON, ruleJSON)
+	}
+
+	resTupleList := make([]*Tuple, 0)
+	for _, rule := range hitRules {
+		tuple := generateTuple(srcPod, dstPod, rule)
+		resTupleList = append(resTupleList, tuple)
+	}
+	// tupleResListJson := make([][]byte, 0)
+	// for _, rule := range resTupleList {
+	// 	ruleJson, err := json.MarshalIndent(rule, "", "  ")
+	// 	if err != nil {
+	// 		log.Fatalf("Error occurred during marshaling. Error: %s", err.Error())
+	// 	}
+	// 	tupleResListJson = append(tupleResListJson, ruleJson)
+	// }
+	return ruleResListJSON, resTupleList, nil
+}
+
+func getCorrespondPod(origin *Input, cacheObj *NPMCache) (*npm.NpmPod, error) {
+	switch origin.Type {
+	case PODNAME:
+		return cacheObj.PodMap[origin.Content], nil
+	case IPADDRS:
+		for _, pod := range cacheObj.PodMap {
+			if pod.PodIP == origin.Content {
+				return pod, nil
+			}
+		}
+		return nil, errInvalidIPAddress
+	case EXTERNAL:
+		return &npm.NpmPod{}, nil
+	default:
+		return nil, errInvalidInput
+	}
+}
+
+// GetInputType returns the type of the input for GetNetworkTuple.
+func GetInputType(input string) InputType {
+	if input == "External" {
+		return EXTERNAL
+	} else if ip := net.ParseIP(input); ip != nil {
+		return IPADDRS
+	} else {
+		return PODNAME
+	}
+}
+
+func generateTuple(src, dst *npm.NpmPod, rule *pb.RuleResponse) *Tuple {
+	tuple := &Tuple{}
+	if rule.Allowed {
+		tuple.RuleType = "ALLOWED"
+	} else {
+		tuple.RuleType = "NOT ALLOWED"
+	}
+	switch rule.Direction {
+	case pb.Direction_EGRESS:
+		tuple.Direction = "EGRESS"
+	case pb.Direction_INGRESS:
+		tuple.Direction = "INGRESS"
+	case pb.Direction_UNDEFINED:
+		// not sure if this is correct
+		tuple.Direction = ANY
+	default:
+		tuple.Direction = ANY
+	}
+	if len(rule.SrcList) == 0 {
+		tuple.SrcIP = ANY
+	} else {
+		tuple.SrcIP = src.PodIP
+	}
+	if rule.SPort != 0 {
+		tuple.SrcPort = strconv.Itoa(int(rule.SPort))
+	} else {
+		tuple.SrcPort = ANY
+	}
+	if len(rule.DstList) == 0 {
+		tuple.DstIP = ANY
+	} else {
+		tuple.DstIP = dst.PodIP
+	}
+	if rule.DPort != 0 {
+		tuple.DstPort = strconv.Itoa(int(rule.DPort))
+	} else {
+		tuple.DstPort = ANY
+	}
+	if rule.Protocol != "" {
+		tuple.Protocol = rule.Protocol
+	} else {
+		tuple.Protocol = ANY
+	}
+	return tuple
+}
+
+func getHitRules(
```

Comment:
This function will become a bottleneck if there are a lot of rules. 
It seems that it needs some optimizations with other data structures.
While we do not implement it in this PR, any cool suggestions (may pre-process while building rules)?

Reply:
I was about to ask you for suggestions 🙂

Reply:
While we do not include these issues in this PR, but just for brainstorming. 
There are two main interesting issues as future works. 
1. How frequently npm analyzer needs to re-build all data structures since iptables and npm cache information are very dynamic and some information in all created data structures frequently become stale. Thus, it needs a good algorithm to partially update or delete created data structure to avoid creating all of them from scratch. 
2. How to (subsequent) query fast.  I think it is related to this function. For them, we may need to have another data structures by using existing data structures information for fast lookup. 

I am not sure how to deal with it now and If you have any thoughts, please share them.  

Leave it as future works. You can contribute them after internship if you want.  This is open-source and your are the main contributor for this part 👍